File: lecture_8/Lecture 8. Source Code/service_tutorial_py/velocity_toggle.py
# Import necessary ROS2 libraries
import rclpy
from rclpy.node import Node
from std_srvs.srv import Empty  # Service type for simple requests
from geometry_msgs.msg import Twist  # Message type for velocity commands

# Define the class for the ROS2 node
class PubVelToggle(Node):

    def __init__(self):
        # Initialize the node with the name 'pubvel_toggle_server'
        super().__init__('pubvel_toggle_server')
        
        # Initialize state variables
        self.forward = True  # Flag to control forward movement
        self.cnt = 0  # Counter for service calls

        # Create a service that listens for requests on the '/toggle_forward' topic
        self.server_ = self.create_service(Empty, '/toggle_forward', self.toggle_forward_callback)
        
        # Create a publisher that sends velocity commands to the 'turtle1/cmd_vel' topic
        self.pub_ = self.create_publisher(Twist, 'turtle1/cmd_vel', 10)
        
        # Create a timer that calls the 'publish_velocity' function every 0.5 seconds
        self.timer_ = self.create_timer(0.5, self.publish_velocity)
        
        # Log a message to indicate that the node has started
        self.get_logger().info("Pubveltoogle now has been started")

    # This function is called when the service receives a request
    def toggle_forward_callback(self, request, response):
        # Increment the counter each time the service is called
        self.cnt += 1
        
        # Toggle the 'forward' flag
        self.forward = not self.forward
        
        # Log the current movement state
        self.get_logger().info(f"Now sending {'forward' if self.forward else 'rotate'} commands.")

        # Return the response
        return response
    
    # This function is called by the timer to publish velocity commands
    def publish_velocity(self):
        # Do nothing if the service has not been called yet
        if self.cnt == 0:
            return

        # Create a new Twist message
        msg = Twist()
        
        # Set the linear and angular velocities based on the 'forward' flag
        msg.linear.x = 1.0 if self.forward else 0.0
        msg.angular.z = 0.0 if self.forward else 1.0

        # Publish the message
        self.pub_.publish(msg)

# The main function where the ROS2 node is executed
def main(args=None):
    # Initialize the ROS2 client library
    rclpy.init(args=args)
    
    # Create an instance of the PubVelToggle node
    node = PubVelToggle()
    
    # Keep the node running until it's interrupted
    rclpy.spin(node)
    
    # Shutdown the ROS2 client library
    rclpy.shutdown()

# This is the entry point of the script
if __name__ == '__main__':
    main()
        

File: lecture_8/src/service_tutorial_cpp/src/velocity_toggle.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_srvs/srv/empty.hpp"
#include "geometry_msgs/msg/twist.hpp"

class PubVelToggle : public rclcpp::Node
{
public:
    PubVelToggle() : Node("pubvel_toggle"), forward_(true), cnt_(0)
    {
        server_ = this->create_service<std_srvs::srv::Empty>(
            "toggle_forward",
            std::bind(&PubVelToggle::toggle_forward_callback, this, std::placeholders::_1, std::placeholders::_2)
        );

        pub_ = this->create_publisher<geometry_msgs::msg::Twist>("turtle1/cmd_vel", 10);

        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(500),
            std::bind(&PubVelToggle::publish_velocity, this)
        );

        RCLCPP_INFO(this->get_logger(), "PubVelToggle node has been started.");
    }

private:
    rclcpp::Service<std_srvs::srv::Empty>::SharedPtr server_;
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr pub_;
    rclcpp::TimerBase::SharedPtr timer_;
    bool forward_;
    int cnt_;

    bool toggle_forward_callback(
        const std::shared_ptr<std_srvs::srv::Empty::Request>,
        std::shared_ptr<std_srvs::srv::Empty::Response>)
    {
        cnt_++;
        forward_ = !forward_;
        RCLCPP_INFO(this->get_logger(), "Now sending %s commands.", forward_ ? "forward" : "rotate");
        return true;
    }

    void publish_velocity()
    {
        if (cnt_ == 0){
            return;
        }

        auto msg = std::make_shared<geometry_msgs::msg::Twist>();

        msg->linear.x = forward_ ? 1.0 : 0.0;
        msg->angular.z = forward_ ? 0.0 : 1.0;

        pub_->publish(*msg);

        RCLCPP_INFO(this->get_logger(), "Published velocity command - linear.x: %.2f, angular.z: %.2f", msg->linear.x, msg->angular.z);
    }
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<PubVelToggle>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}